dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NonterminationDetection.java

```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Option(name = NONTERMINATION_INSTRUMENTATION,
+            description = "(EXPERIMENTAL) " +
+                    "Sets the precision of non-termination detection: only_spinloops, simple, full (default).",
+            secure = true)
+    private Mode mode = Mode.FULL;
+
+    private NonterminationDetection() { }
+
+    public static NonterminationDetection newInstance() {
+        return new NonterminationDetection();
+    }
+
+    public static NonterminationDetection fromConfig(Configuration config) throws InvalidConfigurationException {
+        final NonterminationDetection pass = new NonterminationDetection();
+        config.inject(pass);
+        return pass;
+    }
+
+    @Override
+    public void run(Program program) {
+        if (mode == Mode.ONLY_SPINLOOPS) {
+            // Done by DynamicSpinLoopDetection
+            return;
+        }
+        Preconditions.checkArgument(program.isUnrolled());
+        for (Thread thread : program.getThreads()) {
+            final LoopAnalysis loopAnalysis = LoopAnalysis.onFunction(thread);
+            loopAnalysis.getLoopsOfFunction(thread).forEach(this::instrumentLoop);
+        }
+
+        IdReassignment.newInstance().run(program);
+
+    }
+
+    private void instrumentLoop(LoopAnalysis.LoopInfo loop) {
+        Preconditions.checkArgument(loop.isUnrolled());
+        Preconditions.checkArgument(loop.function() instanceof Thread);
+
+        if (isFullyUnrolled(loop)) {
+            // A fully unrolled loop is always terminating, no need to instrument it.
+            return;
+        }
+
+        final List<LoopAnalysis.LoopIterationInfo> iters;
+        if (mode == Mode.FULL) {
+            iters = loop.iterations();
+        } else if (mode == Mode.SIMPLE) {
+            final int last = loop.iterations().size() - 1;
+            iters = List.of(loop.iterations().get(last - 1));
+        } else {
+            throw new IllegalStateException("Unexpected mode: " + mode);
+        }
+        final Program program = loop.function().getProgram();
+        final BooleanType boolType = TypeFactory.getInstance().getBooleanType();
+        final Label endOfT = (Label) loop.function().getExit();
+        final Event loopHeader = loop.iterations().get(0).getIterationStart();
+
+        for (LoopAnalysis.LoopIterationInfo iter : iters) {
+            final Expression guess = program.newConstant(boolType);
+            final Event nonterm = EventFactory.newJump(guess, endOfT);
+            nonterm.addTags(Tag.NONTERMINATION);
+            nonterm.copyMetadataFrom(loopHeader, SourceLocation.class);
+
+            iter.getIterationEnd().insertAfter(nonterm);
+        }
+
+    }
+
+    private boolean isFullyUnrolled(LoopAnalysis.LoopInfo loop) {
+        final LoopAnalysis.LoopIterationInfo lastIter = loop.iterations().get(loop.iterations().size() - 1);
+        final Event lastEvent = lastIter.getIterationEnd();
+
+        final Event bound = lastEvent.getSuccessor().getSuccessor();
```

Comment:
Does the unrolling guarantee that the bound even is add 2+ after the loop end? If so, document this assumption here

Reply:
I think we guarantee nothing really. While unrolling itself guarantees this, every other pass may break it.
I can add a comment saying that this is a "naive" check. But even if this fails, the code remains correct.

Reply:
I was wrong. The code can accidentally treat a not-fully unrolled loop as fully unrolled and skip its instrumentation. This can result in wrong verdicts for non-termination.